util/util.go

```diff
@@ -283,3 +285,42 @@ func ReadLines(reader *bufio.Reader, count int, maxLineSize int) ([][]byte, erro
 	}
 	return lines, nil
 }
+
+// FieldSlice is the slice of the types.FieldType
+type FieldSlice []*types.FieldType
+
+// CheckTypesCompatibility4PC compares FieldSlice with []*types.FieldType
+// Currently this is only used in plan cache to check whether the types of parameters are compatible.
+// If the types of parameters are compatible, we can use the cached plan.
+func (s FieldSlice) CheckTypesCompatibility4PC(tps []*types.FieldType) bool {
+	if len(s) != len(tps) {
+		return false
+	}
+	for i := range tps {
+		// We only use part of logic of `func (ft *FieldType) Equal(other *FieldType)` here because (1) only numeric and
+		// string types will show up here, and (2) we don't need flen and decimal to be matched exactly to use plan cache
+		tpEqual := (s[i].GetType() == tps[i].GetType()) ||
+			(s[i].GetType() == mysql.TypeVarchar && tps[i].GetType() == mysql.TypeVarString) ||
+			(s[i].GetType() == mysql.TypeVarString && tps[i].GetType() == mysql.TypeVarchar)
+		if !tpEqual || s[i].GetCharset() != tps[i].GetCharset() || s[i].GetCollate() != tps[i].GetCollate() ||
+			(s[i].EvalType() == types.ETInt && mysql.HasUnsignedFlag(s[i].GetFlag()) != mysql.HasUnsignedFlag(tps[i].GetFlag())) {
+			return false
+		}
+		// When the type is decimal, we should compare the Flen and Decimal.
+		// We can only use the plan when both Flen and Decimal should less equal than the cached one.
+		// We assume here that there is no correctness problem when the precision of the parameters is less than the precision of the parameters in the cache.
+		if tpEqual && s[i].GetType() == mysql.TypeNewDecimal && !(s[i].GetFlen() >= tps[i].GetFlen() && s[i].GetDecimal() >= tps[i].GetDecimal()) {
+			return false
+		}
+	}
+	return true
+}
```

Comment:
Better to detach this function from `FieldSlice` and implement it like `checktypescompativility4PC(types1, types2 []*types.FieldType)` and put this new function in `planner/core/plan_cache_utils.go`/